ADAL/src/urlprotocol/ADURLProtocol.m

```diff
@@ -160,108 +173,67 @@ - (void)startLoading
-                            connection:connection
-                              protocol:self];
-    
-    if (!handled)
-    {
-        // Do default handling
-        [challenge.sender performDefaultHandlingForAuthenticationChallenge:challenge];
-        return;
-    }
-    
-    if ([authMethod caseInsensitiveCompare:NSURLAuthenticationMethodNTLM] == NSOrderedSame)
-    {
-        [s_telemetryEvent setNtlm:AD_TELEMETRY_YES];
-    }
-}
-
-#pragma mark - NSURLConnectionDataDelegate Methods
+#pragma mark - NSURLSessionTaskDelegate
 
-- (NSURLRequest *)connection:(NSURLConnection *)connection
-             willSendRequest:(NSURLRequest *)request
-            redirectResponse:(NSURLResponse *)redirectResponse
+- (void)URLSession:(NSURLSession *)session
+              task:(NSURLSessionTask *)task
+willPerformHTTPRedirection:(NSHTTPURLResponse *)response
+        newRequest:(NSURLRequest *)request
+ completionHandler:(void (^)(NSURLRequest * _Nullable))completionHandler
 {
-    (void)connection;
+    (void)session;
     
-    AD_LOG_VERBOSE_F(@"-[ADURLProtocol connection:willSendRequest:]", _correlationId, @"Redirect response: %@. New request:%@", redirectResponse.URL, request.URL);
-    
-    // Disallow HTTP for ADURLProtocol
-    if ([request.URL.scheme isEqualToString:@"http"])
+    if ([request.URL.scheme.lowercaseString isEqualToString:@"http"])
     {
         if ([request.URL.absoluteString.lowercaseString hasPrefix:s_endURL])
         {
             // In this case we want to create an NSURLError so we can intercept the URL in the webview
             // delegate, while still forcing the connection to cancel. This error is the same one the
             // OS sends if it's unable to connect to the host
-            [connection cancel];
-            NSError* failingError = [NSError errorWithDomain:NSURLErrorDomain
+            [task cancel];
+            NSError *failingError = [NSError errorWithDomain:NSURLErrorDomain
                                                         code:-1003
                                                     userInfo:@{ NSURLErrorFailingURLErrorKey : request.URL }];
             [self.client URLProtocol:self didFailWithError:failingError];
         }
-        return nil;
+        completionHandler(nil);
+        return;
     }
     
-    NSMutableURLRequest* mutableRequest = [request mutableCopy];
+    NSMutableURLRequest *mutableRequest = [request mutableCopy];
     
     [ADCustomHeaderHandler applyCustomHeadersTo:mutableRequest];
     [ADURLProtocol addCorrelationId:_correlationId toRequest:mutableRequest];
-
-    if (!redirectResponse)
+    
+    if (!response)
     {
         // If there wasn't a redirect response that means that we're canonicalizing
         // the URL and don't need to cancel the connection or worry about an infinite
         // loop happening so we can just return the response now.
         
-        return mutableRequest;
+        completionHandler(mutableRequest);
```

Comment:
This has a void return type and this is the very last statement. I don't see a need to have a return after this.